### PR TITLE
[PW-3127] Validate payment methods

### PR DESCRIPTION
**Pull request by AlexandrosMor**
**Description**
Validate that paymentMethods response returned to the frontend and see if we have the full response there from adyen and if not then return the full response

**Tested scenarios**

**Fixed issue**:

### CODE REVIEW

File: Helper/ConnectedTerminals.php [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+ * ###### ######  #####( ######  #####. ######  #####          #####  ######
+ * #############  #############  #############  #############  #####  ######
+ *  ############   ############  #############   ############  #####  ######
+ *                                      ######
+ *                               #############
+ *                               ############
+ *
+ * Adyen Payment module (https://www.adyen.com/)
+ *
+ * Copyright (c) 2015 Adyen BV (https://www.adyen.com/)
```
Comment (acampos1916) [inline Helper/ConnectedTerminals.php:18]:
```suggestion
 * Copyright (c) 2020 Adyen BV (https://www.adyen.com/)
```

File: Helper/ConnectedTerminals.php [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+ * See LICENSE.txt for license details.
+ *
+ * Author: Adyen <magento@adyen.com>
+ */
+
+namespace Adyen\Payment\Helper;
+
+
```
Comment (acampos1916) [inline Helper/ConnectedTerminals.php:26]:
```suggestion
```

File: Helper/PaymentMethods.php [hunk 3]
```diff
@@ -165,95 +163,119 @@ protected function fetchAlternativeMethods($country)
...
-        if (isset($responseData['paymentMethods'])) {
-            foreach ($responseData['paymentMethods'] as $paymentMethod) {
+        if (empty($responseData['paymentMethods'])) {
+            return [];
+        }
+
+        $paymentMethods = $responseData['paymentMethods'];
+        $response['paymentMethodsResponse'] = $responseData;
+
+        // Add extra details per payment mmethod
```
Comment (acampos1916) [inline Helper/PaymentMethods.php:210]:
```suggestion
        // Add extra details per payment method
```

File: Helper/PaymentMethods.php [hunk 4]
```diff
@@ -165,95 +163,119 @@ protected function fetchAlternativeMethods($country)
...
+
+            $params = [];
+            $params = array_merge(
+                [
+                    'area' => \Magento\Framework\App\Area::AREA_FRONTEND,
+                    '_secure' => $this->request->isSecure(),
+                    'theme' => $themeCode
+                ],
+                $params
+            );
```
Comment (acampos1916) [inline Helper/PaymentMethods.php:233]: Should we move the logo logic to a separate function?

File: Helper/PaymentMethods.php [hunk 5]
```diff
@@ -165,95 +163,119 @@ protected function fetchAlternativeMethods($country)
...
+                        'url' => $asset->getUrl(),
+                        'width' => $width,
+                        'height' => $height
+                    ];
+                } else {
+                    $icon = [
+                        'url' => 'https://checkoutshopper-live.adyen.com/checkoutshopper/images/logos/medium/' .$paymentMethodCode . '.png',
+                        'width' => 77,
+                        'height' => 50
+                    ];
```
Comment (acampos1916) [inline Helper/PaymentMethods.php:258]: Should we move the logo logic to a separate function?

File: Helper/PaymentMethods.php [hunk 6]
```diff
@@ -165,95 +163,119 @@ protected function fetchAlternativeMethods($country)
...
+            $request["amount"]["currency"] = $currencyCode;
+        }
+
+        $amountValue = $this->adyenHelper->formatAmount(
+            $this->getCurrentPaymentAmount(),
+            !empty($currencyCode) ? $currencyCode : null
+        );
+
+        if (!empty($amountValue)) {
+            $request["amount"]["value"] = $amountValue;
```
Comment (cyattilakiss) [inline Helper/PaymentMethods.php:190]: the amount without the currency I don't think should be added

File: Helper/PaymentMethods.php [hunk 7]
```diff
@@ -165,95 +163,119 @@ protected function fetchAlternativeMethods($country)
...
         ];
 
+        if (!empty($this->getCurrentShopperReference())) {
+            $request["shopperReference"] = $this->getCurrentShopperReference();
+        }
+
+        if (!empty($this->adyenHelper->getCurrentLocaleCode($store->getId()))) {
+            $request["shopperLocale"] = $this->adyenHelper->getCurrentLocaleCode($store->getId());
+        }
+        $currencyCode = $this->getCurrentCurrencyCode($store);
```
Comment (cyattilakiss) [inline Helper/PaymentMethods.php:179]: can the getCurrentCurrencyCode return null?

File: Helper/PaymentMethods.php [hunk 8]
```diff
@@ -165,95 +163,119 @@ protected function fetchAlternativeMethods($country)
...
-            "shopperReference" => $this->getCurrentShopperReference(),
-            "shopperLocale" => $this->adyenHelper->getCurrentLocaleCode($store->getId())
         ];
 
+        if (!empty($this->getCurrentShopperReference())) {
+            $request["shopperReference"] = $this->getCurrentShopperReference();
+        }
+
+        if (!empty($this->adyenHelper->getCurrentLocaleCode($store->getId()))) {
+            $request["shopperLocale"] = $this->adyenHelper->getCurrentLocaleCode($store->getId());
```
Comment (cyattilakiss) [inline Helper/PaymentMethods.php:177]: can the getCurrentLocaleCode return an empty value?

File: Helper/PaymentMethods.php [hunk 9]
```diff
@@ -165,95 +163,119 @@ protected function fetchAlternativeMethods($country)
...
+                        'height' => 50
+                    ];
                 }
-                $paymentMethods[$paymentMethodCode] = $paymentMethod;
+
+                $paymentMethodsExtraDetails[$paymentMethodCode]['icon'] = $icon;
+
+                //todo check if it is needed
+                // check if payment method is an open invoice method
+                $paymentMethodsExtraDetails[$paymentMethodCode]['isOpenInvoice'] =
```
Comment (cyattilakiss) [inline Helper/PaymentMethods.php:265]: keep it here until we get to the frontend maybe we won't need it after refactoring the frontend

File: Helper/PaymentMethods.php (hunk 7 above)
Comment (AlexandrosMor) [reply Helper/PaymentMethods.php:179]: No, if it is empty it will return the base currency

File: Helper/PaymentMethods.php [hunk 10]
```diff
@@ -400,38 +309,111 @@ protected function getCurrentShopperReference()
...
+        $merchantAccount,
+        \Magento\Store\Model\Store $store,
+        $country,
+        \Magento\Quote\Model\Quote $quote
+    ) {
+        $paymentMethodRequest = [
+            "channel" => "Web",
+            "merchantAccount" => $merchantAccount,
+            "countryCode" => $this->getCurrentCountryCode($store, $country),
+            "shopperLocale" => $this->adyenHelper->getCurrentLocaleCode($store->getId())
```
Comment (cyattilakiss) [inline Helper/PaymentMethods.php:329]:
```suggestion
            "shopperLocale" => $this->adyenHelper->getCurrentLocaleCode($store->getId()),
            "amount" => [
                "currency" => $this->getCurrentCurrencyCode($store)
            ]
```

File: Helper/PaymentMethods.php [hunk 11]
```diff
@@ -400,38 +309,111 @@ protected function getCurrentShopperReference()
...
+        ];
 
-        // initialize the adyen client
-        $client = $this->adyenHelper->initializeAdyenClient($storeId, $this->adyenHelper->getPosApiKey($storeId));
+        if (!empty($this->getCurrentShopperReference())) {
+            $paymentMethodRequest["shopperReference"] = $this->getCurrentShopperReference();
+        }
+        $currencyCode = $this->getCurrentCurrencyCode($store);
+        $paymentMethodRequest["amount"]["currency"] = $currencyCode;
```
Comment (cyattilakiss) [inline Helper/PaymentMethods.php:336]:
```suggestion
```

Comment (AlexandrosMor) [reply Helper/PaymentMethods.php:336]: I have to use the  $currencyCode for the format amount:  
```$amountValue = $this->adyenHelper->formatAmount($this->getCurrentPaymentAmount(), $currencyCode);```
That is why I set it later

**Review by cyattilakiss — APPROVED** [review]

**Review by acampos1916 — APPROVED** [review]

**Pull request merged**